File: 프로그래머스/3/133500. 등대/등대.java
import java.util.*;

class Solution {
    
    static ArrayList<ArrayList<Integer>> light = new ArrayList<>();
    static int answer = 0;
    
    public int solution(int n, int[][] lighthouse) {
   
        for(int i=0; i<=n; i++){
            light.add(new ArrayList<>());
        }
        
        for(int i=0; i<lighthouse.length; i++){
            int from = lighthouse[i][0];
            int to = lighthouse[i][1];
            light.get(from).add(to);
            light.get(to).add(from);
        }
        
        dfs(1, 0);
        
        return answer;
    }
    
    public int dfs(int cur, int past){
        
        if(light.get(cur).size() == 1 && light.get(cur).get(0) == past) return 1;
        
        int cnt = 0;
        for(int next : light.get(cur)){
            if(next == past) continue;
            cnt += dfs(next, cur);
        }
        
        if(cnt == 0) return 1;
        answer++;
        return 0;
    }
}

File: 프로그래머스/3/133500. 등대/등대.js
function solution(n, lighthouse) {
    
    const node = Array.from({length:n+1}, () => [])
    const degree = Array(n+1).fill(0)

    for(const [a,b] of lighthouse){
        node[a].push(b)
        node[b].push(a)
        degree[a]++
        degree[b]++
    }

    const que = []

    for(let i=1;i<=n;i++){
        if(degree[i] === 1) que.push(i)
    }

    const visited = Array(n+1).fill(false);
    let answer = 0
    
    while(que.length){
        const now = que.shift();
        if(visited[now]) continue;
        for(const next of node[now]){
            if(visited[next]) continue;
            visited[next] = true;
            answer ++;
            for(const nn of node[next]){
                degree[nn] --;
                if(degree[nn] == 1){
                    que.push(nn);
                }
            }
        }
    }
    
    
    return answer;
}